crates/polars-stream/src/utils/byte_source.rs

```diff
@@ -0,0 +1,329 @@
+    }
+
+    async fn get_size(&self) -> PolarsResult<usize> {
+        match self {
+            Self::Local(v) => v.get_size().await,
+            Self::Cloud(v) => v.get_size().await,
+        }
+    }
+
+    async fn get_range(&self, range: Range<usize>) -> PolarsResult<Bytes> {
+        match self {
+            Self::Local(v) => v.get_range(range).await,
+            Self::Cloud(v) => v.get_range(range).await,
+        }
+    }
+}
+
+impl From<LocalByteSource> for DynByteSource {
+    fn from(value: LocalByteSource) -> Self {
+        Self::Local(value)
+    }
+}
+
+impl From<CloudByteSource> for DynByteSource {
+    fn from(value: CloudByteSource) -> Self {
+        Self::Cloud(value)
+    }
+}
+
+#[derive(Clone)]
+pub enum DynByteSourceBuilder {
+    Local,
+    Cloud,
+}
+
+impl DynByteSourceBuilder {
+    pub async fn try_build_from_path(
+        &self,
+        path: &str,
+        cloud_options: Option<&CloudOptions>,
+    ) -> PolarsResult<DynByteSource> {
+        Ok(match self {
+            Self::Local => LocalByteSource::try_new_from_path(path, cloud_options)
+                .await?
+                .into(),
+            Self::Cloud => CloudByteSource::try_new_from_path(path, cloud_options)
+                .await?
+                .into(),
+        })
+    }
+}
+
+mod tests {
+
+    #[test]
+    fn test() {
+        use polars_io::cloud::CloudOptions;
+        use polars_io::pl_async;
+
+        use super::*;
+        use crate::async_executor;
+
+        // Needs env vars:
+        // * aws_bucket
+        // * aws_access_key_id
+        // * aws_secret_access_key
+        // * aws_region
+
+        let bucket = std::env::var("aws_bucket").unwrap();
+        let cloud_path = format!("s3://{bucket}/nxs/hive_date/date1=2024-01-01/date2=2023-01-01%2000%3A00%3A00.000000/ce0249bc8a1d412f87a7ca87977bd1f2-0.parquet");
+        let cloud_path = cloud_path.as_str();
+
+        let cloud_options = CloudOptions::from_untyped_config(
+            cloud_path,
+            vec![
+                (
+                    "aws_access_key_id",
+                    std::env::var("aws_access_key_id").unwrap().as_str(),
+                ),
+                (
+                    "aws_secret_access_key",
+                    std::env::var("aws_secret_access_key").unwrap().as_str(),
+                ),
+                ("aws_region", std::env::var("aws_region").unwrap().as_str()),
+            ],
+        )
+        .unwrap();
+
+        pl_async::get_runtime().block_on(async {
+            let local_result = read_parquet_metadata(
+                &DynByteSourceBuilder::Local
+                    .try_build_from_path("/Users/nxs/git/polars/.env/iris.parquet", None)
+                    .await
+                    .unwrap(),
+                &async_executor::ExecutorScope::new(),
+            )
+            .await;
+
+            dbg!(local_result.map(|x| x.num_rows));
+
+            let cloud_result = read_parquet_metadata(
```

Comment:
`DynByteSource` lets us use the same `read_parquet_metadata` function to read from both local and cloud